rs/types/management_canister_types/src/lib.rs

```diff
@@ -808,6 +834,106 @@ impl From<pb_canister_state_bits::LogVisibility> for LogVisibility {
+    BoundedVec<MAX_ALLOWED_LOG_VIEWERS_COUNT, UNBOUNDED, UNBOUNDED, PrincipalId>;
+
+/// Log visibility for a canister.
+/// ```text
+/// variant {
+///    controllers;
+///    public;
+///    allowed_viewers: vec principal;
+/// }
+/// ```
+#[derive(Default, Clone, CandidType, Deserialize, Debug, PartialEq, Eq, EnumIter)]
+pub enum LogVisibilityV2 {
+    #[default]
+    #[serde(rename = "controllers")]
+    Controllers,
+    #[serde(rename = "public")]
+    Public,
+    #[serde(rename = "allowed_viewers")]
+    AllowedViewers(BoundedAllowedViewers),
+}
+
+// Ensure backward compatibility with the old `LogVisibility`.
+// First read the new `LogVisibilityV2`, if it fails, try to read the old `LogVisibility`.
+impl Payload<'_> for LogVisibilityV2 {
+    fn decode(blob: &'_ [u8]) -> Result<Self, UserError> {
+        let args = match Decode!([decoder_config()]; blob, Self).map_err(candid_error_to_user_error)
+        {
+            Ok(record) => match record {
+                // TODO(EXC-1670): ignore receiving `AllowedViewers` variant, fallback to default.
+                Self::AllowedViewers(_) => Self::default(),
+                _ => record,
+            },
+            Err(_) => LogVisibility::decode(blob)?.into(),
+        };
+        Ok(args)
+    }
+}
+
+impl From<LogVisibility> for LogVisibilityV2 {
+    fn from(item: LogVisibility) -> Self {
+        match item {
+            LogVisibility::Controllers => Self::Controllers,
+            LogVisibility::Public => Self::Public,
+        }
+    }
+}
+
+impl From<&LogVisibilityV2> for pb_canister_state_bits::LogVisibilityV2 {
+    fn from(item: &LogVisibilityV2) -> Self {
+        use pb_canister_state_bits as pb;
+        match item {
+            LogVisibilityV2::Controllers => pb::LogVisibilityV2 {
+                log_visibility_v2: Some(pb::log_visibility_v2::LogVisibilityV2::Controllers(1)),
+            },
+            LogVisibilityV2::Public => pb::LogVisibilityV2 {
+                log_visibility_v2: Some(pb::log_visibility_v2::LogVisibilityV2::Public(2)),
+            },
+            LogVisibilityV2::AllowedViewers(principals) => pb::LogVisibilityV2 {
+                log_visibility_v2: Some(pb::log_visibility_v2::LogVisibilityV2::AllowedViewers(
+                    pb::LogVisibilityAllowedViewers {
+                        principals: principals
+                            .get()
+                            .iter()
+                            .map(|c| (*c).into())
+                            .collect::<Vec<ic_protobuf::types::v1::PrincipalId>>()
+                            .clone(),
+                    },
+                )),
+            },
+        }
+    }
+}
+
+impl From<pb_canister_state_bits::LogVisibilityV2> for LogVisibilityV2 {
+    fn from(item: pb_canister_state_bits::LogVisibilityV2) -> Self {
+        use pb_canister_state_bits as pb;
+        match item.log_visibility_v2 {
+            None => panic!("Invalid LogVisibility enum variant"),
```

Comment:
I'd return an error instead as per other cases where conversion from proto to Rust might fail and handle it when we deserialize a checkpoint similar to other cases (handling can simply be propagating the error which would result in a panic eventually but it's less surprising I think since it's the usual pattern).

Reply:
Done.